### PR TITLE
[SM-787] Extract authorization from project delete command

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to refactor authorization out of the `DeleteProjectCommand`.

## Code changes

* **bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Projects/ProjectAuthorizationHandler.cs:** 
Adding delete to project authorization handler.

* **bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Projects/DeleteProjectCommand.cs:**
* **src/Core/SecretsManager/Commands/Projects/Interfaces/IDeleteProjectCommand.cs:** 
Extracting authorization from the delete project command.
Removing the extra call to update revision dates in favor of encapsulating the update all in one transaction via `DeleteManyByIdAsync`.

* **bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ProjectRepository.cs:** 
Adding in updating the revision date of secrets associated with the projects being deleted, so this can all be in one database transaction. 

* **bitwarden_license/test/Commercial.Core.Test/SecretsManager/AuthorizationHandlers/Projects/ProjectAuthorizationHandlerTests.cs:** 
Adding in unit test coverage for the delete operation and consolidating update boolean mapping tests.

 * **bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Projects/DeleteProjectCommandTests.cs:** 
Updating unit test coverage for the `DeleteProjectCommand` since authorization is now extracted out.

 * **src/Api/SecretsManager/Controllers/ProjectsController.cs:** 
Updating the bulk delete controller to utilize the project authorization handler.

 * **src/Core/SecretsManager/AuthorizationRequirements/ProjectOperationRequirement.cs:** 
Adding delete as an option for `ProjectOperationRequirement`

 * **test/Api.Test/SecretsManager/Controllers/ProjectsControllerTests.cs:** 
Adding unit test coverage for authorization handler and controller code updates.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by coltonhurst — APPROVED** [review]
This looks great, thanks @Thomas-Avery

**Comment by Thomas-Avery** [comment]
This will get merged into master after https://github.com/bitwarden/server/pull/3021. 

So I've updated this PR to support write access for service accounts on delete. https://github.com/bitwarden/server/pull/2987/commits/67e9f2d30003157e5810159c463c331fa1bf65de

**Review by coltonhurst — APPROVED** [review]
This looks good, thanks @Thomas-Avery

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ProjectRepository.cs [hunk 1]
```diff
@@ -60,16 +60,23 @@ public async Task<IEnumerable<ProjectPermissionDetails>> GetManyByOrganizationId
...
+            foreach (var projectSecret in project.Secrets)
             {
-                dbContext.Remove(project);
-            });
-            await dbContext.SaveChangesAsync();
-        }
+                projectSecret.RevisionDate = utcNow;
+            }
+
+            dbContext.Remove(project);
```
Comment (withinfocus) [inline bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ProjectRepository.cs:76]: 🌱 I wish we had https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#executeupdate-and-executedelete-bulk-updates for this ...

File: src/Api/SecretsManager/Controllers/ProjectsController.cs [hunk 2]
```diff
@@ -127,11 +128,44 @@ public async Task<ProjectResponseModel> GetAsync([FromRoute] Guid id)
...
+        var results = new List<(Project Project, string Error)>();
+
+        foreach (var project in projects)
+        {
+            var authorizationResult =
+                await _authorizationService.AuthorizeAsync(User, project, ProjectOperations.Delete);
+            if (authorizationResult.Succeeded)
+            {
+                projectsToDelete.Add(project);
+                results.Add((project, ""));
```
Comment (withinfocus) [inline src/Api/SecretsManager/Controllers/ProjectsController.cs:158]: ⛏️ I prefer the cleanliness of `string.Empty` vs. the literal here.

File: bitwarden_license/test/Commercial.Core.Test/SecretsManager/Enums/PermissionType.cs [hunk 3]
```diff
@@ -4,5 +4,5 @@ public enum PermissionType
 {
     RunAsAdmin,
     RunAsUserWithPermission,
-    RunAsServiceAccountWithPermission
+    RunAsServiceAccountWithPermission,
```
Comment (withinfocus) [inline bitwarden_license/test/Commercial.Core.Test/SecretsManager/Enums/PermissionType.cs:7]: ⛏️ Errant comma.

**Review by withinfocus — APPROVED** [review]
Minor things not needing change right now. Looks good!

**Pull request merged**